datafusion/physical-plan/src/work_table.rs

```diff
@@ -0,0 +1,193 @@
+/// as is (kind of like a mirror).
+///
+/// Most notably used in the implementation of recursive queries where the
+/// underlying relation does not exist yet but the data will come as the previous
+/// term is evaluated. This table will be used such that the recursive plan
+/// will register a receiver in the task context and this plan will use that
+/// receiver to get the data and stream it back up so that the batches are available
+/// in the next iteration.
+#[derive(Clone, Debug)]
+pub struct WorkTableExec {
+    /// Name of the relation handler
+    name: String,
+    /// The schema of the stream
+    schema: SchemaRef,
+    /// The work table
+    work_table: Arc<WorkTable>,
+    /// Execution metrics
+    metrics: ExecutionPlanMetricsSet,
+}
+
+impl WorkTableExec {
+    /// Create a new execution plan for a worktable exec.
+    pub fn new(name: String, schema: SchemaRef) -> Self {
+        Self {
+            name,
+            schema,
+            metrics: ExecutionPlanMetricsSet::new(),
+            work_table: Arc::new(WorkTable::new()),
+        }
+    }
+
+    pub(super) fn with_work_table(&self, work_table: Arc<WorkTable>) -> Self {
+        Self {
+            name: self.name.clone(),
+            schema: self.schema.clone(),
+            metrics: ExecutionPlanMetricsSet::new(),
+            work_table,
+        }
+    }
+}
+
+impl DisplayAs for WorkTableExec {
+    fn fmt_as(
+        &self,
+        t: DisplayFormatType,
+        f: &mut std::fmt::Formatter,
+    ) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default | DisplayFormatType::Verbose => {
+                write!(f, "WorkTableExec: name={}", self.name)
+            }
+        }
+    }
+}
+
+impl ExecutionPlan for WorkTableExec {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![]
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
+    }
+
+    fn maintains_input_order(&self) -> Vec<bool> {
+        vec![false]
+    }
+
+    fn benefits_from_input_partitioning(&self) -> Vec<bool> {
+        vec![false]
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        None
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(self.clone())
+    }
+
+    /// Stream the batches that were written to the work table.
+    fn execute(
+        &self,
+        partition: usize,
+        _context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream> {
+        // WorkTable streams must be the plan base.
+        if partition != 0 {
+            return Err(DataFusionError::Internal(format!(
+                "WorkTableExec got an invalid partition {} (expected 0)",
+                partition
+            )));
```

Comment:
You can also use the `internal_err!` macro here too if you want

```suggestion
            return internal_err!("WorkTableExec got an invalid partition {partition} (expected 0)");
```

Reply:
done